### PR TITLE
[ENHANCE] Backward Support for Ruby

### PR DESCRIPTION
**Pull request by berniechiu**
### Objective

Since some people still might be using older Ruby version, I guess support for some older ones might be alright @@?

### Features

- Dig module for supporting backward compatibility `dig` like method
- Allow developers to implement its own `dig` without interfering Core classes.

### CODE REVIEW

**Comment by zverok** [comment]
Unfortunately, I don't see it as a way to go. 

The problem is, we extend the core classes, with a pretty core method, implicitly, which can lead to pretty unexpected consequences. A user may already have the implementations for `#dig`, for example — and in the best case, it would be just the port of 2.3's `#dig`, which you also do here. 

But it is completely possible user's `#dig` does something else (or slightly different, like returning `nil` instead of raising). In this case, requiring order of `hm` and user's Hash patch would change the behavior of their code.

So, I see exactly two options to support older Rubies:
1. Specify "2.1" as a minimum Ruby version, go with `backports/2.3.0/{Array,Hash,Struct,OpenStruct}/dig` for _tests_, and document that on 2.1 & 2.2 it is user's responsibility to provide `#dig` method (via the same `backports` for example);
2. On 2.1 & 2.2, conditionally implement "non-patch" dig, like in `lib/hm/dig.rb`:
```ruby
if Hash.respond_to?(:dig) # note: this, not version check
  def dig(what, *keys)
    what.dig(*keys)
  end
else
  def dig(what, *keys)
   ...
  end
end
# ...and so on
```
As all `hm` needs are one level of `dig`, and the whole presence of `dig` method to understand whether it is the "right" type of collection here, an implementation could be simple.

**Comment by berniechiu** [comment]
cool, i've changed it!

**Comment by zverok** [comment]
There is still some misunderstanding here. The main goal is _we should not, ever, change core classes_. So, the two options, to show them in code, work this way (if we are on Ruby 2.1):
```ruby
# Option 1: It is user's reliability to implement #dig
# ========
Hash.instance_method(:dig) # => NameError
require 'hm'
Hash.instance_method(:dig) # => NameError, still!
Hm({a: 1}).dig(:a) # => TypeError: Hash is not diggable
require 'backports/latest' # that what USER did to make Hm work
Hash.instance_method(:dig) # => backports implementation
Hm({a: 1}).dig(:a) # => 1

# Option 2: For 2.1 & 2.2 #dig is implemented internally, but as a module method Hm::Dig.dig(object, *keys)
# =======
Hash.instance_method(:dig) # => NameError
require 'hm'
Hash.instance_method(:dig) # => NameError, still!
Hm({a: 1}).dig(:a) # => 1 -- because internally, Hm recognize it is a hash, and did `Dig.dig(hash, *keys)`
```

**Comment by berniechiu** [comment]
That's very thoughtful, I'll try to remove the dependency in core class

**Comment by berniechiu** [comment]
@zverok all is well 👍

File: lib/hm/algo.rb [hunk 1]
```diff
@@ -87,7 +87,7 @@ def visit_wildcard(what, rest, path, found:, not_found:)
     end
 
     def visit_regular(what, key, rest, path, found:, not_found:) # rubocop:disable Metrics/ParameterLists
-      internal = what.dig(key) or return not_found.(what, [*path, key], rest)
+      internal = Hm::Dig.dig(what, key) or return not_found.(what, [*path, key], rest)
```
Comment (zverok) [inline lib/hm/algo.rb]: `Hm::` is not necessary here.

File: lib/hm/dig.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+class Hm
+  module Dig
+    DIGGABLE_CLASSES = [Hash, Array].freeze
```
Comment (zverok) [inline lib/hm/dig.rb]: Probably `Struct` and `OpenStruct` can be added in future (at least to provide consistent "deep dig"), please just leave a `# TODO:` comment here about that.

File: lib/hm/dig.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+class Hm
+  module Dig
```
Comment (zverok) [inline lib/hm/dig.rb]: Please add `# @private` comment for YARD not show this module in the docs.

File: lib/hm/dig.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    DIGGABLE_CLASSES = [Hash, Array].freeze
+
+    def self.dig(what, *keys)
+      return what.dig(*keys) if what.respond_to?(:dig)
+
+      if diggable?(what)
```
Comment (zverok) [inline lib/hm/dig.rb]: Let's do `diggable?(...) or fail TypeError, ...` here, then there would be less indentation, and less explicit `return`s for the rest of the method :)

**Review by zverok — COMMENTED** [review]
All good now! Left some stylistic comments, and ready to merge after they are addressed.

Thanks for your work and for understanding :+1:

**Comment by berniechiu** [comment]
Nice comments! Resolved now~

**Comment by zverok** [comment]
Thanks! Merged. Releasing new version in a few hours.

**Comment by berniechiu** [comment]
Oh no it failed on Rubocop suggestions lol

**Comment by zverok** [comment]
NP, fixed them at master :)

**Pull request merged**